dask/bytes/s3.py

```diff
@@ -0,0 +1,141 @@
+from __future__ import print_function, division, absolute_import
+
+import logging
+
+from s3fs import S3FileSystem
+
+from .compression import files as compress_files, seekable_files
+
+from ..base import tokenize
+from ..delayed import delayed
+from ..utils import read_block
+
+logger = logging.getLogger(__name__)
+
+
+def read_bytes(path, s3=None, delimiter=None, not_zero=False, blocksize=2**27,
+               sample=True, compression=None, **s3_params):
+    """ Convert location in S3 to a list of delayed values
+
+    Parameters
+    ----------
+    path: string
+        location in S3
+    s3: S3FileSystem (optional)
+    delimiter: bytes
+        An optional delimiter, like ``b'\n'`` on which to split blocks of bytes
+    not_zero: force seek of start-of-file delimiter, discarding header
+    blocksize: int (=128MB)
+        Chunk size
+    sample: bool, int
+        Whether or not to return a sample from the first 10k bytes
+    compression: string or None
+        String like 'gzip' or 'xz'.  Must support efficient random access.
+    **s3_params: keyword arguments
+        Extra keywords to send to boto3 session (anon, key, secret...)
+
+    Returns
+    -------
+    10kB sample header and list of ``dask.Delayed`` objects or list of lists of
+    delayed objects if ``path`` is a globstring.
+    """
+    if compression is not None and compression not in compress_files:
+        raise ValueError("Compression type %s not supported" % compression)
+
+    if s3 is None:
+        s3 = S3FileSystem(**s3_params)
+
+    if '*' in path:
+        filenames = sorted(s3.glob(path))
+        sample, first = read_bytes(filenames[0], s3, delimiter, not_zero,
+                                   blocksize, sample=True,
+                                   compression=compression, **s3_params)
+        rest = [read_bytes(f, s3, delimiter, not_zero, blocksize,
+                       sample=False, compression=compression,  **s3_params)[1]
+                for f in filenames[1:]]
+        return sample, [first] + rest
+    else:
+        if blocksize is None:
+            offsets = [0]
+        else:
+            size = getsize(path, compression, s3)
+            offsets = list(range(0, size, blocksize))
+            if not_zero:
+                offsets[0] = 1
+
+        info = s3.ls(path, detail=True)[0]
+
+        token = tokenize(info['ETag'], delimiter, blocksize, not_zero, compression)
```

Comment:
Nice! We'll get the latest version if the file changed since last access - but remember, s3 caches directory listings, so the ETag may be out of date if passing around the s3 object.